cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1724,9 +1728,9 @@ class ObjectOutputStream final : public io::OutputStream {
 
     RETURN_NOT_OK(EnsureReadyToFlushFromClose());
 
-    auto self = std::dynamic_pointer_cast<ObjectOutputStream>(shared_from_this());
     // Wait for in-progress uploads to finish (if async writes are enabled)
-    return FlushAsync().Then([self]() { return self->FinishPartUploadAfterFlush(); });
+    return FlushAsync().Then(
+        [self = Self()]() { return self->FinishPartUploadAfterFlush(); });
```

Comment:
+1. I proposed the use of `shared_from_this()` in this comment describing issues with `CloseAsync`: https://github.com/apache/arrow/pull/41232#issuecomment-2062514844

Other filesystems implement `CloseAsync` the same way.

Reply:
Rephrasing: it's a good thing that shared_from_this is used here because the default implementation of `CloseAsync` just copies the naked `this` pointer into the `std::function` which can lead to UAFs.

> This change is only a small refactor, it changes nothing semantically.

It's a nice one because you elide one ref-count and don't have to generate code for the destructor of the `self` local variable.

Reply:
Ah, sorry, I might have been overinterpreting your message.